apps/antalmanac/src/stores/ColumnStore.ts

```diff
@@ -0,0 +1,84 @@
+import { create } from 'zustand';
+import analyticsEnum, { logAnalytics } from '$lib/analytics';
+
+/**
+ * Search results are displayed in a tabular format.
+ *
+ * Users can toggle certain columns on/off.
+ */
+export const SECTION_TABLE_COLUMNS = [
+    // These two are omitted since they're not iterated over in the template.
+    // 'scheduleAdd',
+    // 'colorAndDelete',
+    'sectionCode',
+    'sectionDetails',
+    'instructors',
+    'gpa',
+    'dayAndTime',
+    'location',
+    'sectionEnrollment',
+    'restrictions',
+    'status',
+] as const;
+
+export type SectionTableColumn = (typeof SECTION_TABLE_COLUMNS)[number];
+
+interface ColumnStore {
+    // Columns that are enabled. Sometimes columns can be hidden depending on the tab, or perhaps screen size in the future.
+    enabledColumns: boolean[];
+
+    // Columns that are selected in the dropdown
+    selectedColumns: boolean[];
+
+    /**
+     *
+     * @returns The columns that should be displayed in the section table. They need to be both selected and enabled.
+     */
+    getActiveColumns: () => SectionTableColumn[];
+
+    /**
+     * Used by the Select menu in CoursePaneButtonRow to toggle columns on/off.
+     *
+     * @param columns The columns that are selected in the dropdown.
+     */
+    setSelectedColumns: (columns: SectionTableColumn[]) => void;
+
+    /**
+     * Used by TabStore to enable/disable columns depending on the tab.
+     *
+     * @param column The target column.
+     * @param state Whether the column should be enabled or not.
+     */
+    setColumnEnabled: (column: SectionTableColumn, state: boolean) => void;
+}
+
+/**
+ * Store of columns that are currently being displayed in the search results.
+ */
+export const useColumnStore = create<ColumnStore>((set, get) => {
+    return {
+        activeColumns: [...SECTION_TABLE_COLUMNS],
+        enabledColumns: SECTION_TABLE_COLUMNS.map(() => true),
+        selectedColumns: SECTION_TABLE_COLUMNS.map(() => true),
+        getActiveColumns: () =>
+            SECTION_TABLE_COLUMNS.filter((_, index) => get().enabledColumns[index] && get().selectedColumns[index]),
+        setSelectedColumns: (columns: SectionTableColumn[]) => {
+            set(() => {
+                const selectedColumns = SECTION_TABLE_COLUMNS.map((column) => columns.includes(column));
+                return { selectedColumns: selectedColumns };
+            });
+            logAnalytics({
+                category: analyticsEnum.classSearch.title,
+                action: analyticsEnum.classSearch.actions.TOGGLE_COLUMNS,
+            });
+        },
```

Comment:
How come `getActiveColumns` is needed? Could `setSelectedColumns` not set the `selectedColumns` property to this calculated value? Another property could also be added to capture this info if needed.

Otherwise, the best implementation of this particular property would be as a [derived store](https://github.com/pmndrs/zustand/issues/108). It should always be in sync with some other value in the store, so it would be best to either commit to deriving it, or just setting it at the same time as the other values when they change.

Reply:
Isn't this the closest you can get to a derived store? Zustand doesn't seem to natively support it.

Reply:
You could make a property like `currentlyActiveColumns` and then also set that property whenever `setSelectedColumns` or `setColumnEnabled` is called. 

Derived state is automatically updated whenever other state is changed; invoking a function to get the value isn't the same because when multiple components need the derived value, they'll need to individually re-calculate it instead of the value being readily available. It also makes building dependency arrays less clear because functions can't be memo-ized, and it may not be immediately obvious to memo-ize with the two arrays in the store.

Reply:
Yeah I think the double update is the way to go, yeah. I didn't want to do that because it violates single-source of truth, but that's better than forcing the consumer to to update when `setSelectedColumns` or `setColumnEnabled` is called.